### DISCUSSION TITLE
How can I use newest version in U3D?

### DISCUSSION THREAD

**Discussion by sugarcane1990** [bepu/bepuphysics2#233] [Q&A]
I try to use this pretty cool Physics Engine in Unity,But it has some conflict between .net 6.0(bepuphysics2) and .net framework4.7.1(Unity)
In detail，Unity don't contain System.Runtime.Intrinsics.dll and System.Numerics.dll is not the correct version,but when i try to put these from Microsoft.NETCore.App.Ref\6.0.7\ref\net6.0 to Unity pugins folder it do not work.
The error info is:"The name 'BitOperations' does not exist in the current context"(i try to put  ver6.0 System.Numerics.dll in unity plugins folder but it still take this error)
So how can i deal with this problems?

**Comment by george-tsiros** [comment]
according to [MS documentation](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.bitoperations?view=net-6.0)

you want system.runtime.dll

but i _doubt_ it would work

you need to either make unity use net6 or use bepuphysics previous version?

**Reply by sugarcane1990** [reply]
Yes,but Unity don't support for net6.
Or do you konw how to find "BitOperations" source code?If I have this source code,or same replace function/dll,it also can deal with my problems

**Reply by george-tsiros** [reply]
[all the .net 6 source code is online! here on github](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs)!

**Reply by sugarcane1990** [reply]
coooooool！it really helps me a lot!

**Comment by RossNordby** [comment]
Unfortunately, there are a number of incompatibilities between bepuphysics2 and unity at the moment. I would recommend assuming that 2.4 and up are just not going to work until unity's runtimes are updated (check this thread for details: https://forum.unity.com/threads/unity-future-net-development-status.1092205/).

If the goal is to just get anything working at all, you could start with the 2.3.X versions. You'll have an easier time getting things to build with that, but performance will still be extremely bad to the point that the unity built-in options are going to be better.

There are three ways to get bepuphysics2 at full performance in unity:
1. Wait for the runtime updates,
2. Spawn another process to run the simulation (likely going to be incompatible with some kinds of deployment, and annoying overall),
3. Write a C# wrapper for the incomplete and highly experimental C API (https://github.com/RossNordby/scratchpad/blob/main/Abomination/AbominationInterop/BepuPhysicsCPP/BepuPhysics.h), compiled with NativeAOT and treated like a native plugin within unity. (While this would in fact work, please treat this idea with the appropriate level of disrespect for the abomination that it is :P)

**Reply by sugarcane1990** [reply]
thx!got it

**Comment by damian-666** [comment]
I am in general net 6 multiplatform hell as well..  it's not this
particular problem but all sound s very familiar to my last two weeks
 netstandard 2.0 is supposed to be supported but it's No..if u have unity
souce code u might have a chance..also Microsoft made and upgrase-assistant
that almost worked for my tools and I only target desktop DC , gl and
apndroid.  I would wiat this out if u can avoid and proceed with ur own
technology I wish I did not upgrade at this time to net6 library's which
don't work with .net48..  I just didn't want to not know and maybe get
subsumed.


   it will get sorted ..what put me there intha net6 hell is I use monogame
whet fro 3.8 to 3.81.. all my net 48 parts in my level editir can't link to
net 6 libs.  So then Im have to update everything..yes it's faster..lie
20percetn.   So netstand 2.0 is a ghost ..4.8 legacy might work with
windows compatibility introducted by the upgrade assistant tool that makes
quote a mess but might help..  I guess I needed to do all this sooner or
later so I'm going all net6.    I would just revert and wait months for
unity to sort it.    That's why I went with monogame can I have the
source.. and even then it's hell for this transition..but man I got another
20% on my physics module just by building it at net6 not netstandard

**Comment by damian-666** [comment]
also take a look at Silk.Net in net Net foundation git   .. i dont know how
much legacy code you are dealing with  but you might not need Unity..its in
a usable state the the approach they are taking is very interesting, an API
scraper.

**Reply by sugarcane1990** [reply]
im tying to solve the problems that on U3D Android platform how can it deal with more than 300 monster AI by using Fixed-point number and physics engine.
i find Bepuphysics1 has Fixed-point number ver in github,and see Bepuphysics2 as the same time.2 looks like get more good performance,so i just want have a try to use it(or maybe use 2.3ver) on U3D.

**Comment by robsonswiss** [comment]
Hi, can bepuphysics2 not run as standalone and communicate with unity via UDP? It's fairly trivial to setup UDP communication and 500hz can probably achieved. According to unity one can simply run an external physics solution by uninstalling Unity physics and communicating in the fixedupdateloop with the external physics solution.

**Reply by RossNordby** [reply]
Yup, that should work, at least for platforms where it's allowed. I'm not sure what the situation is on mobile.

**Reply by robsonswiss** [reply]
Android, PC, Linux I know all will be fine, not sure about console or IOS. I've tested Unity without any active physics at different fixed update rates and there is performance drop starting at 500hz and more.  Will try and see if I can use Bepuphysics2 via UDP(PC).
